encodeURIComponent should accept any value type

**TypeScript Version:**  2.5.2

**Code**

```ts
encodeURIComponent(true);
```

**Expected behavior:**
tsc compiles with no errors.

**Actual behavior:**
error TS2345: Argument of type 'true' is not assignable to parameter of type 'string'.

**Discussion:**

`encodeURIComponent` [is currently defined](https://github.com/Microsoft/TypeScript/blob/v2.5.2/lib/lib.es6.d.ts#L83) liked this:

```ts
declare function encodeURIComponent(uriComponent: string): string; 
```

In reality, `encodeURIComponent` accepts any value and coerces it to a string if necessary. The first step in [its definition in the ES6 spec](http://www.ecma-international.org/ecma-262/6.0/#sec-encodeuricomponent-uricomponent) reads:

>  1. Let componentString be ToString(uriComponent).

It’s pretty common to pass `true` or `false` or a number like `0` or `10` as a query parameter, for instance. Here are two possible ways to broaden the definition:

```ts
declare function encodeURIComponent(uriComponent: any): string; 

declare function encodeURIComponent(uriComponent: string | number | boolean): string; 
```

The `any` option would be technically correct, but passing `null` or `undefined` or a non-primitive value is usually a bug (in my experience). You can use your own judgment about how broadly the `uriComponent` argument should be defined. I’m specifically interested in using `number` and `boolean`.

--- a/src/lib/es5.d.ts
+++ b/src/lib/es5.d.ts
@@ -60,7 +60,7 @@ declare function encodeURI(uri: string): string;
   * Encodes a text string as a valid component of a Uniform Resource Identifier (URI).
   * @param uriComponent A value representing an encoded URI component.
   */
-declare function encodeURIComponent(uriComponent: string): string;
+declare function encodeURIComponent(uriComponent: string | number | boolean): string;
 
 /**
   * Computes a new string in which certain characters have been replaced by a hexadecimal escape sequence.

--- a/tests/baselines/reference/1.0lib-noErrors.js
+++ b/tests/baselines/reference/1.0lib-noErrors.js
@@ -78,7 +78,7 @@ declare function encodeURI(uri: string): string;
   * Encodes a text string as a valid component of a Uniform Resource Identifier (URI).
   * @param uriComponent A value representing an encoded URI component.
   */
-declare function encodeURIComponent(uriComponent: string): string;
+declare function encodeURIComponent(uriComponent: string | number | boolean): string;
 
 interface PropertyDescriptor {
     configurable?: boolean;
@@ -1141,7 +1141,7 @@ declare var Array: {
     <T>(...items: T[]): T[];
     isArray(arg: any): boolean;
     prototype: Array<any>;
-}
+}
 
 //// [1.0lib-noErrors.js]
 /* *****************************************************************************

--- a/tests/baselines/reference/1.0lib-noErrors.symbols
+++ b/tests/baselines/reference/1.0lib-noErrors.symbols
@@ -98,12 +98,12 @@ declare function encodeURI(uri: string): string;
   * Encodes a text string as a valid component of a Uniform Resource Identifier (URI).
   * @param uriComponent A value representing an encoded URI component.
   */
-declare function encodeURIComponent(uriComponent: string): string;
+declare function encodeURIComponent(uriComponent: string | number | boolean): string;
 >encodeURIComponent : Symbol(encodeURIComponent, Decl(1.0lib-noErrors.ts, 73, 48))
 >uriComponent : Symbol(uriComponent, Decl(1.0lib-noErrors.ts, 79, 36))
 
 interface PropertyDescriptor {
->PropertyDescriptor : Symbol(PropertyDescriptor, Decl(1.0lib-noErrors.ts, 79, 66))
+>PropertyDescriptor : Symbol(PropertyDescriptor, Decl(1.0lib-noErrors.ts, 79, 85))
 
     configurable?: boolean;
 >configurable : Symbol(PropertyDescriptor.configurable, Decl(1.0lib-noErrors.ts, 81, 30))
@@ -130,7 +130,7 @@ interface PropertyDescriptorMap {
 
     [s: string]: PropertyDescriptor;
 >s : Symbol(s, Decl(1.0lib-noErrors.ts, 91, 5))
->PropertyDescriptor : Symbol(PropertyDescriptor, Decl(1.0lib-noErrors.ts, 79, 66))
+>PropertyDescriptor : Symbol(PropertyDescriptor, Decl(1.0lib-noErrors.ts, 79, 85))
 }
 
 interface Object {
@@ -217,7 +217,7 @@ declare var Object: {
 >getOwnPropertyDescriptor : Symbol(getOwnPropertyDescriptor, Decl(1.0lib-noErrors.ts, 141, 32))
 >o : Symbol(o, Decl(1.0lib-noErrors.ts, 149, 29))
 >p : Symbol(p, Decl(1.0lib-noErrors.ts, 149, 36))
->PropertyDescriptor : Symbol(PropertyDescriptor, Decl(1.0lib-noErrors.ts, 79, 66))
+>PropertyDescriptor : Symbol(PropertyDescriptor, Decl(1.0lib-noErrors.ts, 79, 85))
 
     /** 
       * Returns the names of the own properties of an object. The own properties of an object are those that are defined directly 
@@ -250,7 +250,7 @@ declare var Object: {
 >o : Symbol(o, Decl(1.0lib-noErrors.ts, 171, 19))
 >p : Symbol(p, Decl(1.0lib-noErrors.ts, 171, 26))
 >attributes : Symbol(attributes, Decl(1.0lib-noErrors.ts, 171, 37))
->PropertyDescriptor : Symbol(PropertyDescriptor, Decl(1.0lib-noErrors.ts, 79, 66))
+>PropertyDescriptor : Symbol(PropertyDescriptor, Decl(1.0lib-noErrors.ts, 79, 85))
 
     /**
       * Adds one or more properties to an object, and/or modifies attributes of existing properties. 
@@ -2084,3 +2084,4 @@ declare var Array: {
 >prototype : Symbol(prototype, Decl(1.0lib-noErrors.ts, 1140, 31))
 >Array : Symbol(Array, Decl(1.0lib-noErrors.ts, 973, 23), Decl(1.0lib-noErrors.ts, 1133, 11))
 }
+

--- a/tests/baselines/reference/1.0lib-noErrors.types
+++ b/tests/baselines/reference/1.0lib-noErrors.types
@@ -98,9 +98,9 @@ declare function encodeURI(uri: string): string;
   * Encodes a text string as a valid component of a Uniform Resource Identifier (URI).
   * @param uriComponent A value representing an encoded URI component.
   */
-declare function encodeURIComponent(uriComponent: string): string;
->encodeURIComponent : (uriComponent: string) => string
->uriComponent : string
+declare function encodeURIComponent(uriComponent: string | number | boolean): string;
+>encodeURIComponent : (uriComponent: string | number | boolean) => string
+>uriComponent : string | number | boolean
 
 interface PropertyDescriptor {
     configurable?: boolean;
@@ -1911,3 +1911,4 @@ declare var Array: {
     prototype: Array<any>;
 >prototype : any[]
 }
+

--- a/tests/baselines/reference/parserRegularExpression4.types
+++ b/tests/baselines/reference/parserRegularExpression4.types
@@ -61,7 +61,7 @@ if (Ca.test(c.href) || Ba.test(c.href) && /(\\?|&)adurl=/.test(c.href) && !/(\\?
 >"&q=" + encodeURIComponent(W("q") || W("as_q") || A) : string
 >"&q=" : "&q="
 >encodeURIComponent(W("q") || W("as_q") || A) : string
->encodeURIComponent : (uriComponent: string) => string
+>encodeURIComponent : (uriComponent: string | number | boolean) => string
 >W("q") || W("as_q") || A : any
 >W("q") || W("as_q") : any
 >W("q") : any

--- a/tests/cases/conformance/decorators/1.0lib-noErrors.ts
+++ b/tests/cases/conformance/decorators/1.0lib-noErrors.ts
@@ -78,7 +78,7 @@ declare function encodeURI(uri: string): string;
   * Encodes a text string as a valid component of a Uniform Resource Identifier (URI).
   * @param uriComponent A value representing an encoded URI component.
   */
-declare function encodeURIComponent(uriComponent: string): string;
+declare function encodeURIComponent(uriComponent: string | number | boolean): string;
 
 interface PropertyDescriptor {
     configurable?: boolean;
@@ -1141,4 +1141,4 @@ declare var Array: {
     <T>(...items: T[]): T[];
     isArray(arg: any): boolean;
     prototype: Array<any>;
-}
+}

--- a/tests/lib/lib.d.ts
+++ b/tests/lib/lib.d.ts
@@ -77,7 +77,7 @@ declare function encodeURI(uri: string): string;
   * Encodes a text string as a valid component of a Uniform Resource Identifier (URI).
   * @param uriComponent A value representing an encoded URI component.
   */
-declare function encodeURIComponent(uriComponent: string): string;
+declare function encodeURIComponent(uriComponent: string | number | boolean): string;
 
 /**
   * Computes a new string in which certain characters have been replaced by a hexadecimal escape sequence.

